### PR TITLE
[system-probe] Count registers correctly

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Counting registers (additions) to the conntrack cache was not being done correctly as it did not account for replacing values and evictions.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details that should be tested during the QA.

### CODE REVIEW

File: pkg/network/netlink/conntracker.go [hunk 1]
```diff
@@ -232,14 +232,16 @@ func (ctr *realConntracker) register(c Con) int {
...
 	ctr.Lock()
 	defer ctr.Unlock()
 
+	l := ctr.cache.Len()
+	defer func() {
+		atomic.AddInt64(&ctr.stats.registers, int64(ctr.cache.Len()-l))
+		atomic.AddInt64(&ctr.stats.registersTotalTime, time.Now().UnixNano()-then)
+	}()
+
 	ctr.cache.Add(c, true)
```
Comment (leeavital) [inline pkg/network/netlink/conntracker.go:245]: Does this need to be done in a defer at all? This function has almost no branching, and it doesn't really improve readability to handle stats in a defer.

File: pkg/network/netlink/conntracker.go [hunk 2]
```diff
@@ -220,6 +221,8 @@ func (ctr *realConntracker) loadInitialState(events <-chan Event) {
 			ctr.cache.Add(c, false)
 		}
 	}
+
+	atomic.AddInt64(&ctr.stats.registers, int64(ctr.cache.Len()-l))
```
Comment (brycekahle) [inline pkg/network/netlink/conntracker.go:225]: Could this be incorrect, if something else modifies `ctr.cache` in another goroutine?

**Comment by brycekahle** [comment]
I think there is value in knowing both:
1. How many cache operations are being done, regardless of their result on the overall size of the cache.
2. The change in size of the cache over time.

It was my understanding that `registers` was essentially cache `Add` operations, rather than knowing that a new item was added. Could we simply add another metric that is the size of the cache?

File: pkg/network/netlink/conntracker.go (hunk 2 above)
Comment (hmahmood) [reply pkg/network/netlink/conntracker.go:225]: During the initial load the assumption is there is only one go-routine.

**Comment by hmahmood** [comment]
> I think there is value in knowing both:
> 
> 1. How many cache operations are being done, regardless of their result on the overall size of the cache.
> 2. The change in size of the cache over time.
> 
> It was my understanding that `registers` was essentially cache `Add` operations, rather than knowing that a new item was added. Could we simply add another metric that is the size of the cache?

1. With the non-LRU cache, `registers` vs `unregisters` were essentially symmetric. Continuing to tie them to operations on the cache, with it being an LRU, means they're not going to be symmetric anymore. In practice I see registers outpace unregisters. I am fine with still tracking it the old way though.
2. We already have a stat for tracking the cache size, `system_probe.conntrack.state_size`. Did you have something else in mind?

**Comment by brycekahle** [comment]
> 1. In practice I see registers outpace unregisters. I am fine with still tracking it the old way though

Should we track evictions as a separate metric, that way we have full information about how items left the cache?

**Comment by brycekahle** [comment]
> 2\. Did you have something else in mind?

Nope. That works!

**Comment by hmahmood** [comment]
> > 1. In practice I see registers outpace unregisters. I am fine with still tracking it the old way though
> 
> Should we track evictions as a separate metric, that way we have full information about how items left the cache?

Sure, done

File: pkg/network/netlink/conntracker.go [hunk 3]
```diff
@@ -160,19 +161,28 @@ func (ctr *realConntracker) GetStats() map[string]int64 {
...
+		m["gets_total"] = gets
+		m["nanoseconds_per_get"] = getTimeTotal / gets
 	}
-	if ctr.stats.registers != 0 {
-		m["registers_total"] = ctr.stats.registers
-		m["registers_dropped"] = ctr.stats.registersDropped
-		m["nanoseconds_per_register"] = ctr.stats.registersTotalTime / ctr.stats.registers
+
+	registers := atomic.LoadInt64(&ctr.stats.registers)
+	if atomic.LoadInt64(&ctr.stats.registers) != 0 {
```
Comment (brycekahle) [inline pkg/network/netlink/conntracker.go:172]: Should this be `registers != 0`?

Comment (hmahmood) [reply pkg/network/netlink/conntracker.go]: Yes, fixed

**Review by brycekahle — APPROVED** [review]

**Pull request merged**